### ISSUE TITLE
Memory Issues with running Daily CNN example script

### ISSUE THREAD

**Issue by Georgepu1** [CarperAI/trlx#307]
### 🚀 The feature, motivation, and pitch

Hi, I was wondering if anyone ran into memory issues with running the T5 script (https://github.com/CarperAI/trlx/blob/main/examples/summarize_daily_cnn/t5_summarize_daily_cnn.py), and if there were any recommendations on optimizing memory consumption?

It seems like the model is replicated across GPUs, and that increasing the number of GPUs does not resolve the memory consumption (e.g. selecting 1 process vs 4/8 processes still result in each machine using ~20-25gb of memory). Is it possible to split the model across the processes/GPUs?

Thanks!

### Alternatives

For example, gradient checkpointing could be used or using huggingface's device_map = "auto" when defining the model (e.g. AutoModelForSeq2SeqLM.from_pretrained(model_name, device_map="auto"))

**Comment by ZHAOTING** [comment]
IIRC, the current huggingface transformers do not seem to support model parallelism (i.e. split the model across multiple GPUs).

But trlx uses accelerate, which can enable deepspeed usage. With deepspeed, there are several efficient ways to reduce GPU memory use:
1) [ZeRO optimization](https://www.deepspeed.ai/tutorials/zero/) (at three different stages);
2) mixed precision (i.e. fp16 and bf16);
3) other options such as CPU offload and activation checkpointing.

Notice that 
* ZeRO works for multiple GPUs.
* Mixed precision has a tradeoff between accuracy vs. memory+speed, while others have a tradeoff between memory vs. time.

**Comment by PhungVanDuy** [comment]
> ### 🚀 The feature, motivation, and pitch
> Hi, I was wondering if anyone ran into memory issues with running the T5 script (https://github.com/CarperAI/trlx/blob/main/examples/summarize_daily_cnn/t5_summarize_daily_cnn.py), and if there were any recommendations on optimizing memory consumption?
> 
> It seems like the model is replicated across GPUs, and that increasing the number of GPUs does not resolve the memory consumption (e.g. selecting 1 process vs 4/8 processes still result in each machine using ~20-25gb of memory). Is it possible to split the model across the processes/GPUs?
> 
> Thanks!
> 
> ### Alternatives
> For example, gradient checkpointing could be used or using huggingface's device_map = "auto" when defining the model (e.g. AutoModelForSeq2SeqLM.from_pretrained(model_name, device_map="auto"))
> 
> ### Additional context
> _No response_

I can make sure it is able to work on 40GB GPUs, I did not try to run it on smaller GPUs. I suggest that you can use `bf16` (T5 does not work well with fp16 in my experiments so far). You can also try to use ZeRO2 offload with deepspeed configuration, or reduce batch size at this [line](https://github.com/CarperAI/trlx/blob/main/examples/summarize_daily_cnn/configs/ppo_config_cnn_daily.yml#L5).

**Comment by Georgepu1** [comment]
Thanks all for the great recommendations! I think when i tried running with fp16, I got overflow/underflow errors, but I can look into other DeepSpeed optimizations. 

My main concern was that the T5 daily_cnn script already uses 20-30GB memory, and if we wanted to add support for another reward model, it would OOM before training even starts. 

Have either of you tried building a full end-to-end RLHF pipeline with the T5 model (similar to using GPT-j in the /examples/summarize_rlhf directory) with having an actual reward model in the pipeline and not the METEOR metric?

**Comment by PhungVanDuy** [comment]
> Thanks all for the great recommendations! I think when I tried running with fp16, I got overflow/underflow errors, but I can look into other DeepSpeed optimizations.
> 
> My main concern was that the T5 daily_cnn script already uses 20-30GB memory, and if we wanted to add support for another reward model, it would OOM before training even starts.
> 
> Have either of you tried building a full end-to-end RLHF pipeline with the T5 model (similar to using GPT-j in the /examples/summarize_rlhf directory) with having an actual reward model in the pipeline and not the METEOR metric?

I strongly suggest you use bf16 or fp32 with small batch size for T5 instead of fp16. 
Thank you for the suggestion to run RLHF with T5, with a setting for summarization, if want build RLHF with T5 you still need to use a GPT family model for the reward model where you can put [prompt] [output] as input for the model but T5 it not straight forward when input only able prompt and output should be output at the decoder. That is our concern to apply RLHF on T5. I still think about how to design a reward model with T5, but that will be great if you can use our current reward model (GPT-J) and use T5 to train on top of it. Thank you.

**Comment by Georgepu1** [comment]
Yeah that makes sense why making a RM with T5 would be a little tricky, I can look into it in the coming weeks and let you know if there's anything promising design-wise.

I'll probably try retraining the GPT-j reward model (e.g. using GPT-neo 1.3B) to see if I can run the RM with the [prompt] + [output] pair into the model for the scalar value. Still running into OOM issues with GPT-j with DeepSpeed optimization (e.g. offloading params/optim) and bf16 even with an A100 instance, but I'll keep you updated.

I'll leave this issue open for another day if anyone else has suggestions, but this was incredibly useful. 

Thank you so much!

GPT-Neo 1.3B

**Comment by LouisCastricato** [comment]
I've used T5 as a reward model before, it was fine. I ran into zero problems. I asked T5 if A > B or B > A, and then computed ELO from that (which I then normalized.) It was a bit slow but not too awful. You can, of course, also just add a CLS token to T5 which you then map to a scalar. But in that case you're only really using the T5 encoder (which to be clear is quite a good encoder, I've also used that as a standalone RM.)
